### PR TITLE
feat(analytics, ui): event based analytics service plus tailwind UI for the settings

### PR DESCRIPTION
**Pull request by connoratrug**
### Analytics
Allow maintainer (admin) user to setup analytics events. When the enduser of the web interface and executes a certain action ( for example clicks on the 'contact' button) a signal is send to a analytics service ( third party; site-improve, google, .. or emx2 analytics endpoint). The app maintainer (admin ? )should be able to configure which events are send in response to what ui element the user interacts with ( initial 'click' event only). The maintainer should be able to configure these events on a per schema basis. 

What are the main changes you did:
- Add 'Trigger' model to the System schema ( a trigger hold the configuration as defined by the maintainer )
- Create a client module that fetches the triggers configured for a schema and initialises event triggers in the client application ( if necessary it fetches code to handle the triggers, i.e. send the analytics events for a certain ( set of ) provider(s).
- Create a API to c.r.u.d triggers
- Create a ui to  c.r.u.d triggers

### Tailwind (analytics) ui ( pre alpha)
Replace bootstrap emx apps with single tailwind+nuxt app , ssr set to false and ui components though layers ( form tailwind-components module)

What are the main changes you did:
- explain what you changed and essential considerations.

how to test:
- this pr only configures the analytics module the ssr-catalogue app, to test the service using this app 
   - Navigate to the tailwind pre alpa ui -> /apps/ui/#/ 
   - Signin ( if not signed in )
   - Select the schema that contains the tables for the catalogue app 
   - Select 'Analytics' in the main menu
   - Press the 'Add' button
   - Add a new trigger, give the trigger a name , copy a valid css selector ( from the page in the ssr-catalogue app ) into the cssSelector field.
   - Visit the page that contains the selector and 'click' the page element that was setup by the cssSelector
   - You can inspect the console log and the network tab to check the setup has run and the event has been fired.

   - If you have acces to the configured analytics provider ( currently only site-improve ) login to the system and check the event have received ( note: this only works if correct server and analytic-key has been configure in site-improve. you may not have the rights the do so ) 

todo:
- [ ] updated docs in case of new feature
- [ ] added/updated tests
- [ ] added/updated testplan to include a test for this fix, including ref to bug using # notation

### CODE REVIEW

**Comment by mswertz** [comment]
why don't you use the standard forms for the crud? Just like we do with the 'tasks'?

(ideally of course the tailwind version of that).

File: apps/ui/pages/analytics.vue [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+<template>
+  <Container>
+    <PageHeader title="Analytics" />
+    <ContentBlock
+      class="mt-1"
+      title="Triggers"
+      description="Trigger that fire analytics events"
+    >
+      <Table>
```
Comment (mswertz) [inline apps/ui/pages/analytics.vue:65]: This is code I would want to use higher order components. I don't want to maintain copies of the code for these kinds of screens multiple times.

In the bootstrap we had 'MolgenisTable' for the table that already has the edit buttons.  
If you must you can use seperate 'RowButtonAdd/Delete/Edit' for the edit dialogues.

File: backend/molgenis-emx2-analytics/src/main/java/org/molgenis/emx2/analytics/repository/TriggerRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    AtomicInteger inserted = new AtomicInteger();
+
+    this.database.tx(
+        db -> {
+          Schema sysSchema = db.getSchema(SYSTEM_SCHEMA);
+          if (sysSchema == null) {
+            throw new MolgenisException("failed to add analytics Trigger");
+          }
+          Table triggerTable = sysSchema.getTable(TRIGGER_TABLE_NAME);
+          Row triggerRow =
```
Comment (mswertz) [inline backend/molgenis-emx2-analytics/src/main/java/org/molgenis/emx2/analytics/repository/TriggerRepositoryImpl.java:65]: interestingly I have been playing with idea to extend emx2 Table class to also accept `add(Object)` where the backend would automatically map records to rows.

File: backend/molgenis-emx2-analytics/src/main/java/org/molgenis/emx2/analytics/repository/TriggerRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+              row(
+                  NAME,
+                  trigger.name(),
+                  CSS_SELECTOR,
+                  trigger.cssSelector(),
+                  APP_NAME,
+                  trigger.appName(),
+                  SCHEMA_NAME,
+                  trigger.schemaName());
+          inserted.set(triggerTable.insert(triggerRow));
```
Comment (mswertz) [inline backend/molgenis-emx2-analytics/src/main/java/org/molgenis/emx2/analytics/repository/TriggerRepositoryImpl.java:75]: why don't you insert all these in a batch?

**Review by mswertz — CHANGES_REQUESTED** [review]
I vote 'request changes' because of 'metaStore' and massive code duplication frontend for the crud.

In bootstrap I would use the higher order components. I strongly feel we should do that here too? Btw I might not entirely understand the forms you are making so that might change my mind.

File: apps/tailwind-components/pages/DataFetch.other.vue [hunk 4]
```diff
@@ -0,0 +1,123 @@
+<script setup lang="ts">
```
Comment (mswertz) [inline apps/tailwind-components/pages/DataFetch.other.vue:1]: ok, I might like this over the client as a pattern but there is much knowledge in the client that we are now duplicating.

File: apps/tailwind-components/plugins/floating-vue.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { defineNuxtPlugin } from "#app";
+import FloatingVue from "floating-vue";
+import "floating-vue/dist/style.css";
+
+export default defineNuxtPlugin(({ vueApp }) => {
+  vueApp.use(FloatingVue);
```
Comment (mswertz) [inline apps/tailwind-components/plugins/floating-vue.ts:6]: would be nice to explain what this does so we understand this plugin

File: apps/tailwind-components/server/routes/[schema]/api/trigger.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { joinURL } from "ufo";
+export default defineEventHandler((event) => {
+ const config = useRuntimeConfig(event); 
+  console.log("proxy trigger request : ", event.path);
+  const schema = getRouterParam(event, "schema") || "";
+  const triggerPreviewLocation = "https://preview-emx2-pr-3946.dev.molgenis.org/"
```
Comment (mswertz) [inline apps/tailwind-components/server/routes/[schema]/api/trigger.ts:6]: huh?

File: apps/tailwind-components/server/routes/[schema]/api/trigger.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  //const location = joinURL(config.public.apiBase, schema, "api", "trigger")
+  const location = joinURL(triggerPreviewLocation, schema, "api", "trigger")
+  console.log("to : ", location);
```
Comment (mswertz) [inline apps/tailwind-components/server/routes/[schema]/api/trigger.ts:9]: ?

File: apps/ui/server/routes/[schema]/api/trigger/[trigger].ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { joinURL } from "ufo";
+export default defineEventHandler((event) => {
+ const config = useRuntimeConfig(event); 
+  console.log("proxy single trigger request : ", event.path, ' type: ', event.method);
+  const location = joinURL(config.public.apiBase, event.path)
+  console.log("to : ", location);
+  return proxyRequest(event, location);
```
Comment (mswertz) [inline apps/ui/server/routes/[schema]/api/trigger/[trigger].ts:7]: I thought this is bad practice to do these proxy? Should redirect maybe?

**Review by mswertz — CHANGES_REQUESTED** [review]

**Comment by mswertz** [comment]
> > I vote 'request changes' because of 'metaStore' and massive code duplication frontend for the crud.
> > In bootstrap I would use the higher order components. I strongly feel we should do that here too? Btw I might not entirely understand the forms you are making so that might change my mind.
> 
> could the be a bit more specific on the issues and requested changes ? imo there in only code duplication until we delete molgenis-components module

You should have split this PR.

**Comment by mswertz** [comment]
ok, first course of action is to extract all changes related to tailwind-components

**Comment by MaxPostema** [comment]
note: tailwind-components are extracted to a separate pr: #4077

File: apps/emx2-analytics/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+Goal of the module is to allow service maintainers to track certain user actions. 
+For example maintainer of a Data Catalogue application might what to know when and by who a 'Contact' button is clicked.
+
+This module when installed sets up 'Triggers' attched to page elements ( for example a Button in a given location).
+Besides the need to install and configure this module in a app ( or compoment ),
+the triggers ( what events to run when ) should be created using the analytics api or web ui.
+
+### Install the module 
+
+use package manager to install the module '@molgenis/emx2-analytics'
```
[the remark this answers is not in the document]
Comment (mswertz) [reply apps/emx2-analytics/README.md:12]: we agree not to publish actually.

File: apps/emx2-analytics/package.json [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  },
+  "devDependencies": {
+    "@vitejs/plugin-vue": "^5.0.5",
+    "prettier": "2.8.8",
+    "typescript": "^5.2.2",
+    "vite": "^5.3.1",
+    "vite-plugin-dts": "^3.9.1",
+    "vue-tsc": "^2.0.21"
+  },
+  "publishConfig": {
```
[the remark this answers is not in the document]
Comment (mswertz) [reply apps/emx2-analytics/package.json:26]: config can stay as long we don't call it.

File: apps/emx2-analytics/tsconfig.app.json [hunk 11]
```diff
@@ -0,0 +1,27 @@
+{
+  "compilerOptions": {
```
[the remark this answers is not in the document]
Comment (mswertz) [reply apps/emx2-analytics/tsconfig.app.json:2]: other files also have this.

File: apps/emx2-analytics/tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,35 @@
+{
+  "compilerOptions": {
```
[the remark this answers is not in the document]
Comment (mswertz) [reply apps/emx2-analytics/tsconfig.json:2]: one for the module and one for the app? we will check

File: apps/nuxt3-ssr/nuxt.config.ts [hunk 13]
```diff
@@ -14,7 +14,7 @@ export default defineNuxtConfig({
       emx2Theme: "",
       emx2Logo: "",
       siteTitle: "MOLGENIS",
-      analyticsKey: "",
+      analyticsKey: "1466456",
```
Comment (mswertz) [inline apps/nuxt3-ssr/nuxt.config.ts:17]: empty

**Review by mswertz — APPROVED** [review]

**Pull request merged**